scapy/layers/ipsec.py

```diff
@@ -127,8 +130,57 @@ class ESP(Packet):
 
 bind_layers(IP, ESP, proto=socket.IPPROTO_ESP)
 bind_layers(IPv6, ESP, nh=socket.IPPROTO_ESP)
-bind_layers(UDP, ESP, dport=4500)  # NAT-Traversal encapsulation
-bind_layers(UDP, ESP, sport=4500)  # NAT-Traversal encapsulation
+
+
+class UDPENCAP(Packet):
```

Comment:
Now that I think about it, it would make much more sense to split this packet in 2 packets.
Then you can use a `dispatch_hook` your select between then on dissection, and let the user pick one on build.
This makes sense for me, what do you think?

I dislike the implementation of `UFPENCAP`. Having a conditionalfield on a class attribute isn't what we usually do in Scapy, and all in all what you did looks too complicated.

Reply:
I am grateful for any suggestion from you to make the implementation simpler, because I've been struggling with the three cases from RFC 3948 section 2 for a while now.  But I'm not sure I completely understand what you are proposing.

According to [RFC 3948, section 2](https://www.rfc-editor.org/rfc/rfc3948#section-2), there are three cases of packet formats:
```
Ether / IP / UDP / ESP                 # UDP-encapsulated ESP message  (the default)
Ether / IP / UDP / 0x00000000 / IKE    # UDP-encapsulated IKE message  (indicated by four byte non-esp marker)
Ether / IP / UDP / 0xFF                # NAT-keepalive  (indicated by a single 'magic' byte)
```

#### Current implementation: single packet UDPENCAP

I admit that the *implementation* of UDPENCAP is not the simplest possible, but it was designed with the goal to make its *usage* as simple as possible, by handling the three packet format cases automatically: All the user needs to do is to place the UDPENCAP packet after the UDP packet and then add the ESP, IKE, or no packet:

```
Ether() / IP() / UDP() / UDPENCAP() / ESP()
Ether() / IP() / UDP() / UDPENCAP() / IKE()
Ether() / IP() / UDP() / UDPENCAP()
```

The UDPENCAP class automatically takes care of the fill bytes depending on the class which follows, via the conditional fields `non_esp` and `nat_keepalive`. And, more important, in all three cases building and dissecting are inverse, i.e. `Ether(raw(pkt))` has the same packet structure (see the test cases.)

#### Alternative implementation: two packets NON_ESP, NAT_KEEPALIVE

Now lets assume that I introduce two packets `NON_ESP` and `NAT_KEEPALIVE` for cases two and three. (In the first case, there is actually no need for an extra packet). This would be simpler to implement, but now the user would be responsible for choosing the correct combination when building, because only the following three combinations are valid:
```
Ether() / IP() / UDP() / ESP()
Ether() / IP() / UDP() / NON_ESP() / IKE()
Ether() / IP() / UDP() / NAT_KEEPALIVE()
```
Any other combination would be invalid and would not dissect to the same combination again. (In most cases, `Ether(raw(pkt))` will simply fail.)

```
Ether() / IP() / UDP() / NON_ESP() / ESP()
Ether() / IP() / UDP() / IKE()
Ether() / IP() / UDP() / NAT_KEEPALIVE() / ESP()
Ether() / IP() / UDP() / NAT_KEEPALIVE() / IKE()
```

In the case of my UDPENCP implementation, the user could also build invalid combinations, but only by explicitly changing the default values of the `non_esp` resp. `nat_keepalive` fields.

https://github.com/secdev/scapy/blob/15a330e831f77f77387295762c4ede4f5f0d346e/scapy/layers/ipsec.py#L145-L152

Is the second variant what you had in mind, even though it is not as simple to use as the first one?

Reply:
> This makes sense for me, what do you think?

I understand that the automatism contained in UDENCAP might contradict Scapy's principle of leaving the user complete control over what packet he/she builds, even if it results in complete garbage. So if you prefer variant two, I'll make the necessary changes.

If I misunderstood you and you had a different solution in mind, please let me know.

Reply:
You summed it up really well, I do mean option 2 !

Thanks a lot for your patience on this 😛

Reply:
Thanks a lot in turn for your patience with me 😅

Reply:
> Then you can use a `dispatch_hook` your select between then on dissection

Which class would you choose for the dispatch hook (and bind to port 4500)?  I could bind ESP directly, since it is the default case

```python
bind_bottom_up(UDP, ESP, dport=4500)
bind_bottom_up(UDP, ESP, sport=4500)
bind_layers(UDP, ESP, dport=4500, sport=4500)
```

but then the ESP.dispatch_hook would have to take into account whether the underlayer is UDP or not (i.e, distinguish between UDP encapsulated ESP vs. raw ESP).

Alternatively, I could leave UDPENCAP as 'abstract' class which only implements the dispatch_hook and never appears after dissection.

https://github.com/secdev/scapy/blob/15a330e831f77f77387295762c4ede4f5f0d346e/scapy/layers/ipsec.py#L180-L182

Reply:
The first variant (using ESP.dispatch_hook) has again some implicit automatism: What if the user wants to bind ESP directly to UDP on port 4711 without the multiplexing of RFC 3948?

Reply:
Done. I hope you dislike my new implementation less  :wink: